Change Aircraft.IsMoving to be true only when moving horizontally

reaperrr opened the pull request:
This gets rid of the overly simplistic "above ground = moving" assumption aircraft currently make, removing the need to work around it in other places.

In case we need a vertical movement check later, I added `IsMovingVertically` as a precaution.

Prerequisite for #11886.

reaperrr committed:
Move Aircraft.firstTick above constructor

reaperrr committed:
Aircraft.ReserveSpawnBuilding style fixes

reaperrr committed:
Change Aircraft.IsMoving to only return true when moving horizontally

Additionally added IsMovingVertically for potential edge cases where vertical movement needs to be considered.

reaperrr committed:
Remove work-around for aircraft from WithMoveAnimation

pchote commented on OpenRA.Mods.Common/Traits/Render/WithMoveAnimation.cs:
```diff
@@ -43,8 +43,7 @@ public void Tick(Actor self)
 			var oldCachedPosition = cachedPosition;
 			cachedPosition = self.CenterPosition;
```
This is now unused and can be removed.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -266,12 +276,12 @@ public Actor GetActorBelow()
-			/* HACK: not spawning in the air, so try to assoc. with our afld. */
-			var afld = GetActorBelow();
-			if (afld == null)
+			// HACK: Not spawning in the air, so try to associate with our spawner.
```
Idea for a future PR: check for `ParentActorInit` in the constructor and drop this hack.

GraionDilach commented:
This will break voxel helicopters - good example is the RA2 Blackhawk.

reaperrr commented:
> This will break voxel helicopters - good example is the RA2 Blackhawk.

How so?

pchote commented:
It looks like they are using `WithVoxelWalkerBody` for it, which seems like a horrible kludge.

GraionDilach commented:
Only `WithVoxelWalkerBody` allows for animated voxels. All the other voxel traits take the first hva frame of the voxel and keep it static.

reaperrr commented:
> Only `WithVoxelWalkerBody` allows for animated voxels. All the other voxel traits take the first hva frame of the voxel and keep it static.

Then make a copy/derivative of `WithVoxelWalkerBody`, name it `WithVoxelHelicopterBody`, adapt it to check for DAT > 0 instead of movement.

pchote commented:
Or add a more generic animated voxel trait and use upgrades to turn it on and off.

abcdefg30 commented:
Why aren't you using `IsMovingVertically`?

reaperrr commented:
> Why aren't you using `IsMovingVertically`?

Assuming that was directed at me, that would make the TD Orca - and presumably TS aircraft voxels later - pitch forward when launching vertically from the helipad. Considering vertical-only movement as 'IMove.IsMoving' would cause more problems than it would solve.

That being said, once this is merged I could add a bool property to #11886 that allows modders to apply upgrades on vertical movement as well, if they want to (then `IsMovingVertically` would no longer be unused, either).

abcdefg30 commented:
> That being said, once this is merged I could add a bool property to #11886 that allows modders to apply upgrades on vertical movement as well, if they want to (then  IsMovingVertically  would no longer be unused, either).

That sounds good to me.

reaperrr committed:
Move `IsMovingVertically` to IMove

- avoids an Aircraft trait look-up in #11886
- potentially useful for future additional locomotors (jumpjet), custom locomotors in external mods, possibly future updates to default ground locomotor (Mobile) with regard to behavior on slopes

reaperrr commented:
Updated, moved `IsMovingVertically` down a level to `IMove`, so it can be used by other locomotors and traits (and saves us an `Aircraft` look-up in #11886).

On that note, @pchote do you think splitting that into `IsMovingUp` and `IsMovingDown` could have any merit with regard to movement on slopes (TS had speed modifiers for moving up/down a slope, for example)?

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

pchote reviewed (approved):
:+1:

pchote commented:
This is just an internal refactoring, so doesn't need a changelog entry.
